src/circuit/gadget/ecc.rs

```diff
@@ -1,140 +1,336 @@
+    type ScalarFixedShort: Clone + Debug;
     /// Variable representing an elliptic curve point.
-    type Point: Clone + fmt::Debug;
+    type Point: Clone + Debug;
+    /// Variable representing the x-coordinate of an elliptic curve point.
+    type X: Clone + Debug;
     /// Variable representing the set of fixed bases in the circuit.
-    type FixedPoints: FixedPoints<C>;
+    type FixedPoints: Clone + Debug;
     /// Variable representing a fixed elliptic curve point (constant in the circuit).
-    type FixedPoint: Clone + fmt::Debug;
+    type FixedPoint: Clone + Debug;
+    /// Variable representing a fixed elliptic curve point (constant in the circuit)
+    /// to be used in scalar multiplication with a short signed exponent.
+    type FixedPointShort: Clone + Debug;
+
+    /// Witnesses the given base field element as a private input to the circuit for variable-base scalar mul.
+    fn witness_scalar_var(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Base>,
+    ) -> Result<Self::ScalarVar, Error>;
 
-    /// Witnesses the given scalar as a private input to the circuit.
-    fn witness_scalar(
-        layouter: &mut impl Layouter<Self>,
+    /// Witnesses the given full-width scalar as a private input to the circuit for fixed-base scalar mul.
+    fn witness_scalar_fixed(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C::Scalar>,
-    ) -> Result<Self::Scalar, Error>;
+    ) -> Result<Self::ScalarFixed, Error>;
+
+    /// Witnesses the given signed short scalar as a private input to the circuit for fixed-base scalar mul.
+    fn witness_scalar_fixed_short(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Scalar>,
+    ) -> Result<Self::ScalarFixedShort, Error>;
 
     /// Witnesses the given point as a private input to the circuit.
     fn witness_point(
-        layouter: &mut impl Layouter<Self>,
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C>,
     ) -> Result<Self::Point, Error>;
 
-    /// Gets a fixed point into the circuit.
-    fn get_fixed(
-        layouter: &mut impl Layouter<Self>,
+    /// Extracts the x-coordinate of a point.
+    fn extract_p(point: &Self::Point) -> &Self::X;
+
+    /// Returns a fixed point that had been previously loaded into the circuit.
+    /// The pre-loaded cells are used to set up equality constraints in other
+    /// parts of the circuit where the fixed base is used.
+    fn get_fixed(&self, fixed_points: Self::FixedPoints) -> Result<Self::FixedPoint, Error>;
+
+    /// Returns a fixed point to be used in scalar multiplication with a signed
+    /// short exponent.
+    fn get_fixed_short(
+        &self,
         fixed_points: Self::FixedPoints,
-    ) -> Result<Self::FixedPoint, Error>;
+    ) -> Result<Self::FixedPointShort, Error>;
```

Comment:
Is the intent that all fixed bases can be fetched in either full-width or short form? That makes sense generically, but given that fixed bases are loaded into the circuit ahead of time, that implies we pay the cost of loading all of them in both full and short form (which could be fine if the short form can just be a subset of the full form). If this is the case, please document this on `EccInstructions::FixedPoints`. If not, then I guess we also need an `EccInstructions::FixedPointsShort`.

Reply:
The short form is not a subset of the full form because the last `u`- and `z`- values are different. We load `ValueCommitV` as a short fixed base, and the other four bases as full-width. I've added `EccInstructions::FixedPointsShort`.